### PR TITLE
Removing global seeds

### PR DESCRIPTION
**Pull request by omardrwch**
The global seeding of `rlberry` cannot handle well parallel threads or processes and may lead to unexpected behavior (e.g. #42).

This PR removes the global seeding to make `rlberry` more compatible with multi-threading/processing. This is done by introducing a `Seeder` class that wraps a numpy SeedSequence, and can be used to seed different agents and environments. Then, seeding works as in the following example:

```python
from rlberry.seeding import Seeder

seeder = Seeder(123)

# Each Seeder instance has a random number generator (rng)
# See https://numpy.org/doc/stable/reference/random/generator.html to check the
# methods available in rng.
seeder.rng.integers(5)
seeder.rng.normal()
print(type(seeder.rng))
# etc

# Environments and agents should be seeded using a single seeder,
# to ensure that their random number generators are independent.
from rlberry.envs import gym_make
from rlberry.agents import RSUCBVIAgent
env = gym_make('MountainCar-v0')
env.reseed(seeder)

agent = RSUCBVIAgent(env)
agent.reseed(seeder)


# Environments and Agents have their own seeder and rng!
print("env seeder: ", env.seeder)
print("random sample from env rng: ", env.rng.normal())
print("agent seeder: ", agent.seeder)
print("random sample from agent rng: ", agent.rng.normal())


# A seeder can spawn other seeders that are independent from it
# This is useful to seed two different threads, using seeder1
# in the first thread, and seeder2 in the second thread.
seeder1, seeder2 = seeder.spawn(2)
```

Also:

* `Seeder` instances are available by default in the `Agent` and `Model` classes. We have access to their random number generators in an `rng` attribute: `agent.rng` and `env.rng`. 
* `AgentStats` has now a `seed` parameter in its constructor, and handles automatically the seeding of all its environments and agents.
* `rlberry.seeding` provides a function `set_external_seed` to seed external libraries, such as torch. This function is called in each thread/process of `AgentStats` to ensure reproducibility.

### CODE REVIEW

File: examples/demo_from_stable_baselines.py [hunk 1]
```diff
@@ -73,6 +70,15 @@ def policy(self, observation, **kwargs):
...
+    #
+    @classmethod
+    def sample_parameters(cls, trial):
+        learning_rate = trial.suggest_loguniform('learning_rate', 1e-5, 1)
+
+        return {'learning_rate': learning_rate}
+
 
 #
 # Traning one agent
```
Comment (xuedong) [inline examples/demo_from_stable_baselines.py:84]: Typo herre

File: examples/demo_from_stable_baselines_atari.py [hunk 2]
```diff
@@ -76,6 +78,15 @@ def policy(self, observation, **kwargs):
...
+    #
+    @classmethod
+    def sample_parameters(cls, trial):
+        learning_rate = trial.suggest_loguniform('learning_rate', 1e-5, 1)
+
+        return {'learning_rate': learning_rate}
+
 
 #
 # Traning one agent
```
Comment (xuedong) [inline examples/demo_from_stable_baselines_atari.py:92]: Same typo here

File: examples/demo_twinrooms.py [hunk 3]
```diff
@@ -5,8 +5,10 @@
 
 env = TwinRooms()
 env = DiscretizeStateWrapper(env, n_bins=20)
+env.reseed(41)
```
Comment (xuedong) [inline examples/demo_twinrooms.py:8]: Probably replace with a seeder here later

File: examples/demo_seeding.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+from rlberry.envs import gym_make
+from rlberry.agents import RSUCBVIAgent
+env = gym_make('MountainCar-v0')
+env.reseed(seeder)
+
+agent = RSUCBVIAgent(env)
+agent.reseed(seeder)
+
+
+# Environments and Agents have their own seeder and rng!
```
Comment (xuedong) [inline examples/demo_seeding.py:24]: I'm not sure the message of the following lines is clear enough

File: rlberry/seeding/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from .seeder import Seeder
+from .seeding import safe_reseed
+from .seeding import set_external_seed
```
Comment (xuedong) [inline rlberry/seeding/__init__.py:3]: No new line at the end of file

File: rlberry/seeding/seeder.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        ----------
+        seed_seq : np.random.SeedSequence, rlberry.seeding.Seeder or int, default : None
+            Seed sequence from which to spawn the random number generator.
+            If None, generate random seed.
+            If int, use as entropy for SeedSequence.
+            If seeder, use seeder.seed_seq
+        spawn_seed_seq : bool, default : True
+            If True, uses seed_seq to spawn a new seed sequence (strongly recommended) for the Seeder.
+            If False, uses the input seed_seq to define the Seeder.
+            Warning: Setting to false can lead to unexpect behavior. This argument is only used internally
```
Comment (xuedong) [inline rlberry/seeding/seeder.py:34]: unexpect -> unexpected

**Review by xuedong — COMMENTED** [review]
Only a few small changes to be made, I didn't find big issues.
